File: internal/design/preview/canvas.go
package preview

import (
	"bytes"
	_ "embed"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"sync"
	"time"
)

// The canvas is the read-only counterpart of the single-artifact preview: one
// pan-and-zoom surface holding every artifact of a session as an artboard, so a
// user watching an agent design something sees the work appear and change
// without having to open a tab per artifact.
//
// It is deliberately view-only. The artboards are live documents in iframes, but
// a transparent capture layer sits over each one: the user pans, zooms and
// reads, and never edits by hand. Editing stays the agent's job, which is what
// keeps the file on disk the single source of truth.
//
// Like an artifact grant, a canvas grant is a capability: the token in the path
// is the only way in, it is bound to a session, and it expires.

// CanvasSegment is the token-free first segment of every canvas URL. It cannot
// collide with an artifact token because tokens are hex and this is not.
const CanvasSegment = "_canvas"

// CanvasPath is the route prefix the canvas owns.
const CanvasPath = Prefix + CanvasSegment + "/"

// CanvasGrant is one canvas published for viewing.
type CanvasGrant struct {
	Token     string    `json:"token"`
	SessionID string    `json:"session_id,omitempty"`
	ExpiresAt time.Time `json:"expires_at"`
}

// Artboard is one artifact as the canvas sees it. The preview package knows
// nothing about the design model, so the whole struct is filled in by the
// [Options.Artboards] provider that internal/design installs.
type Artboard struct {
	ID    string `json:"id"`
	Title string `json:"title"`
	Slug  string `json:"slug,omitempty"`
	Kind  string `json:"kind,omitempty"`
	// URL is the address of the artifact document the artboard frames.
	URL string `json:"url"`
	// Width and Height are the logical viewport the artboard is rendered at,
	// before the canvas zoom scales it. A deck is 1280x720, a page 1440x900.
	Width  int `json:"width"`
	Height int `json:"height"`
	// Version is the artifact's snapshot number, shown on the artboard label.
	Version int `json:"version,omitempty"`
	// Status is "ready", "building" or "error". The canvas renders a badge for
	// anything that is not ready.
	Status string `json:"status,omitempty"`
	// Note carries the error message, or whatever the status needs to explain.
	Note string `json:"note,omitempty"`
	// Revision changes whenever the document underneath changes. The canvas
	// watches it to flash the artboard and to force a reload of its frame.
	Revision  uint64    `json:"revision"`
	UpdatedAt time.Time `json:"updated_at,omitempty"`
}

// canvasState is the payload the page polls.
type canvasState struct {
	Artboards []Artboard `json:"artboards"`
	// ServedAt lets the page show when it last heard from the server, which is
	// the difference between "nothing is happening" and "the poll is broken".
	ServedAt time.Time `json:"served_at"`
	Error    string    `json:"error,omitempty"`
}

//go:embed canvas.html
var canvasPage []byte

//go:embed canvas.js
var canvasScript []byte

// canvasRegistry is the canvas half of the grant table. It is a separate map
// from the artifact grants because a canvas has no directory and serves no
// files: sharing one map would mean every file route had to ask which kind of
// grant it just resolved.
type canvasRegistry struct {
	mu     sync.RWMutex
	grants map[string]*CanvasGrant // token -> grant
	// bySession keeps one canvas per session, so re-presenting the canvas
	// returns the URL already open in the user's browser instead of a second
	// window looking at the same thing.
	bySession map[string]string
}

func newCanvasRegistry() *canvasRegistry {
	return &canvasRegistry{
		grants:    make(map[string]*CanvasGrant),
		bySession: make(map[string]string),
	}
}

// PublishCanvas mints (or refreshes) the canvas grant of a session and returns
// it. Sessions without an id share the process-wide canvas, which is what a
// plain CLI invocation wants.
func (s *Server) PublishCanvas(sessionID string) (CanvasGrant, error) {
	if err := s.checkAccess(); err != nil {
		return CanvasGrant{}, err
	}
	reg := s.canvas
	reg.mu.Lock()
	defer reg.mu.Unlock()

	token, ok := reg.bySession[sessionID]
	if !ok {
		var err error
		if token, err = newToken(); err != nil {
			return CanvasGrant{}, err
		}
		reg.bySession[sessionID] = token
	}
	grant := &CanvasGrant{
		Token:     token,
		SessionID: sessionID,
		ExpiresAt: time.Now().Add(s.opts.TTL),
	}
	reg.grants[token] = grant
	reg.sweepLocked()
	return *grant, nil
}

// CanvasURL returns the address of a published canvas.
func (s *Server) CanvasURL(sessionID string) (string, error) {
	reg := s.canvas
	reg.mu.RLock()
	token, ok := reg.bySession[sessionID]
	grant := reg.grants[token]
	reg.mu.RUnlock()
	if !ok || grant == nil {
		return "", errors.New("preview: no canvas is published for this session")
	}
	u := &url.URL{Path: CanvasPath + token + "/"}
	return s.base() + u.String(), nil
}

// RevokeCanvas drops the canvas grant of a session.
func (s *Server) RevokeCanvas(sessionID string) {
	reg := s.canvas
	reg.mu.Lock()
	defer reg.mu.Unlock()
	if token, ok := reg.bySession[sessionID]; ok {
		delete(reg.grants, token)
		delete(reg.bySession, sessionID)
	}
}

func (r *canvasRegistry) sweepLocked() {
	now := time.Now()
	for token, grant := range r.grants {
		if now.After(grant.ExpiresAt) {
			delete(r.grants, token)
			delete(r.bySession, grant.SessionID)
		}
	}
}

func (r *canvasRegistry) resolve(token string) (*CanvasGrant, bool) {
	r.mu.RLock()
	grant, ok := r.grants[token]
	r.mu.RUnlock()
	if !ok || time.Now().After(grant.ExpiresAt) {
		return nil, false
	}
	return grant, true
}

// serveCanvas handles everything under CanvasPath. The routes are:
//
//	/preview/_canvas/{token}/            the page
//	/preview/_canvas/{token}/canvas.js   its script
//	/preview/_canvas/{token}/artboards   the JSON the page polls
//
// The script sits inside the token space rather than beside the bridge because
// it is only ever loaded by a page that already holds a token, and keeping it
// there means the canvas adds no token-free route to the server.
func (s *Server) serveCanvas(w http.ResponseWriter, r *http.Request) {
	rest := strings.TrimPrefix(r.URL.Path, CanvasPath)
	token, rest, _ := strings.Cut(rest, "/")
	grant, ok := s.canvas.resolve(token)
	if !ok {
		// Same silence an unknown artifact token gets: a wrong token must not
		// tell the holder that some other canvas exists.
		http.NotFound(w, r)
		return
	}

	s.writeSecurityHeaders(w)
	w.Header().Set("Cache-Control", "no-store, must-revalidate")

	switch rest {
	case "", "index.html":
		w.Header().Set("Content-Type", "text/html; charset=utf-8")
		http.ServeContent(w, r, "canvas.html", time.Time{}, bytes.NewReader(canvasPage))
	case "canvas.js":
		w.Header().Set("Content-Type", "application/javascript; charset=utf-8")
		http.ServeContent(w, r, "canvas.js", time.Time{}, bytes.NewReader(canvasScript))
	case "artboards":
		s.serveArtboards(w, grant)
	default:
		http.NotFound(w, r)
	}
}

func (s *Server) serveArtboards(w http.ResponseWriter, grant *CanvasGrant) {
	state := canvasState{ServedAt: time.Now().UTC()}
	if s.opts.Artboards == nil {
		state.Error = "this build serves no artboards"
	} else if boards, err := s.opts.Artboards(grant.SessionID); err != nil {
		state.Error = err.Error()
	} else {
		state.Artboards = boards
	}
	// A stable order keeps an artboard from jumping across the canvas between
	// two polls, which would make the surface unusable while an agent works.
	sort.SliceStable(state.Artboards, func(i, j int) bool {
		return state.Artboards[i].ID < state.Artboards[j].ID
	})
	if state.Artboards == nil {
		state.Artboards = []Artboard{}
	}
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	if err := json.NewEncoder(w).Encode(state); err != nil {
		// The response is already committed; there is nothing to say to the
		// client, and the page will simply keep the artboards it has.
		_ = fmt.Errorf("preview: encode canvas state: %w", err)
	}
}

File: internal/design/preview/canvas.js
// The design canvas: every artifact of a session as an artboard on one
// pan-and-zoom surface, refreshed while the agent works.
//
// It is a viewer, never an editor. Each artboard frames the very document the
// preview server serves, so what the user sees is the file on disk — but a
// transparent capture layer takes every pointer event, so the only things the
// user can do are look, pan, zoom and focus. Editing stays with the agent.
//
// State arrives by polling `artboards` rather than over a socket: the payload is
// a handful of rows, the page must work in both the mounted and the loopback
// deployment, and each artboard's own live-reload script already refreshes its
// content. The poll is here for the canvas *layout* — boards appearing,
// disappearing, changing size or version.

(function () {
  'use strict';

  var POLL_MS = 1000;
  var GAP = 96;                 // space between artboards, in canvas units
  var LABEL_H = 26;             // height of an artboard's label strip
  var ROW_MAX = 4600;           // wrap a row once it passes this width
  var MIN_ZOOM = 0.02;
  var MAX_ZOOM = 4;

  var viewport = document.getElementById('viewport');
  var world = document.getElementById('world');
  var zoomLabel = document.getElementById('zoom');
  var jump = document.getElementById('jump');
  var followBtn = document.getElementById('follow');
  var dot = document.getElementById('dot');
  var liveText = document.getElementById('live-text');
  var empty = document.getElementById('empty');
  var activityList = document.getElementById('activity-list');

  // view is the canvas transform: world coordinates -> screen.
  var view = { x: 0, y: 0, k: 1 };
  var boards = new Map();       // id -> {data, el, iframe, box}
  var follow = true;
  var userPlacedView = false;   // true once the user pans or zooms by hand
  var failures = 0;

  // ---------------------------------------------------------------- render

  function applyView() {
    world.style.transform = 'translate(' + view.x + 'px,' + view.y + 'px) scale(' + view.k + ')';
    zoomLabel.textContent = Math.round(view.k * 100) + '%';
  }

  function boardSize(data) {
    var w = data.width > 0 ? data.width : 1440;
    var h = data.height > 0 ? data.height : 900;
    return { w: w, h: h };
  }

  // layout packs the artboards into wrapping rows, largest row height driving
  // the next row's offset. It is deterministic: the same list always lands in
  // the same place, so an artboard never jumps while the user is reading it.
  function layout(list) {
    var x = 0, y = 0, rowH = 0;
    list.forEach(function (data) {
      var size = boardSize(data);
      if (x > 0 && x + size.w > ROW_MAX) {
        x = 0;
        y += rowH + GAP + LABEL_H;
        rowH = 0;
      }
      data._x = x;
      data._y = y;
      x += size.w + GAP;
      rowH = Math.max(rowH, size.h);
    });
  }

  function worldBounds() {
    var minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
    boards.forEach(function (entry) {
      var size = boardSize(entry.data);
      minX = Math.min(minX, entry.data._x);
      minY = Math.min(minY, entry.data._y - LABEL_H);
      maxX = Math.max(maxX, entry.data._x + size.w);
      maxY = Math.max(maxY, entry.data._y + size.h);
    });
    if (minX === Infinity) return null;
    return { x: minX, y: minY, w: maxX - minX, h: maxY - minY };
  }

  function fitBox(box, padding) {
    if (!box) return;
    var pad = padding === undefined ? 80 : padding;
    var vw = viewport.clientWidth - pad * 2;
    var vh = viewport.clientHeight - pad * 2;
    if (vw <= 0 || vh <= 0) return;
    var k = Math.min(vw / box.w, vh / box.h);
    view.k = Math.max(MIN_ZOOM, Math.min(MAX_ZOOM, k));
    view.x = pad + (vw - box.w * view.k) / 2 - box.x * view.k;
    view.y = pad + (vh - box.h * view.k) / 2 - box.y * view.k;
    applyView();
  }

  function fitAll() {
    fitBox(worldBounds());
  }

  function focusBoard(id, padding) {
    var entry = boards.get(id);
    if (!entry) return;
    var size = boardSize(entry.data);
    fitBox({ x: entry.data._x, y: entry.data._y - LABEL_H, w: size.w, h: size.h + LABEL_H }, padding);
  }

  // ---------------------------------------------------------------- boards

  function buildBoard(data) {
    var el = document.createElement('section');
    el.className = 'board';

    var label = document.createElement('div');
    label.className = 'label';

    var name = document.createElement('span');
    name.className = 'name';
    label.appendChild(name);

    var meta = document.createElement('span');
    meta.className = 'meta';
    label.appendChild(meta);

    var badge = document.createElement('span');
    badge.className = 'badge';
    badge.hidden = true;
    label.appendChild(badge);

    var open = document.createElement('button');
    open.className = 'open';
    open.textContent = 'open ↗';
    open.title = 'Open this artifact on its own, in a new tab';
    label.appendChild(open);

    var frame = document.createElement('div');
    frame.className = 'frame';

    var iframe = document.createElement('iframe');
    iframe.setAttribute('loading', 'lazy');
    iframe.setAttribute('title', data.title || data.id);
    // No sandbox attribute, deliberately: the framed document is same-origin
    // so its injected live-reload poll can reach the token-scoped `_live`
    // endpoint, which is what makes an artboard refresh itself the moment the
    // agent rewrites the file. The artifact CSP is what confines it, and the
    // capture layer above is what keeps the user from editing it.
    frame.appendChild(iframe);

    var capture = document.createElement('div');
    capture.className = 'capture';
    frame.appendChild(capture);

    var fail = document.createElement('div');
    fail.className = 'fail';
    fail.hidden = true;
    frame.appendChild(fail);

    el.appendChild(label);
    el.appendChild(frame);
    world.appendChild(el);

    var entry = {
      data: data, el: el, iframe: iframe, name: name, meta: meta,
      badge: badge, fail: fail, open: open, loaded: ''
    };
    open.addEventListener('click', function (event) {
      event.stopPropagation();
      if (entry.data.url) window.open(entry.data.url, '_blank', 'noopener');
    });
    // Double-clicking anywhere on the board frames it. The capture layer is
    // above the document, so this works without the click ever reaching it.
    el.addEventListener('dblclick', function () {
      userPlacedView = true;
      focusBoard(entry.data.id, 40);
    });
    return entry;
  }

  function paintBoard(entry) {
    var data = entry.data;
    var size = boardSize(data);
    entry.el.style.left = data._x + 'px';
    entry.el.style.top = (data._y - LABEL_H) + 'px';
    entry.el.style.width = size.w + 'px';

    entry.name.textContent = data.title || data.slug || data.id;
    var bits = [];
    if (data.kind) bits.push(data.kind);
    bits.push(size.w + '×' + size.h);
    if (data.version > 0) bits.push('v' + data.version);
    entry.meta.textContent = bits.join(' · ');

    var status = data.status || 'ready';
    if (status === 'ready') {
      entry.badge.hidden = true;
    } else {
      entry.badge.hidden = false;
      entry.badge.className = 'badge ' + status;
      entry.badge.textContent = status === 'building' ? 'building…' : status;
      entry.badge.title = data.note || '';
    }

    entry.iframe.style.width = size.w + 'px';
    entry.iframe.style.height = size.h + 'px';

    if (!data.url) {
      entry.fail.hidden = false;
      entry.fail.textContent = data.note || 'This artifact has not been rendered yet.';
      entry.iframe.hidden = true;
      return;
    }
    entry.fail.hidden = true;
    entry.iframe.hidden = false;

    // Reload only when the document actually changed. Reloading on every poll
    // would restart animations and scroll positions once a second.
    var want = data.url + '#rev=' + (data.revision || 0);
    if (entry.loaded !== want) {
      entry.loaded = want;
      entry.iframe.src = data.url;
    }
  }

  function flash(entry) {
    entry.el.classList.add('flash');
    window.setTimeout(function () { entry.el.classList.remove('flash'); }, 900);
  }

  // ---------------------------------------------------------------- state

  function sameShape(a, b) {
    return a.title === b.title && a.status === b.status && a.version === b.version &&
      a.width === b.width && a.height === b.height && a.url === b.url &&
      a.revision === b.revision && a.note === b.note;
  }

  function apply(state) {
    var list = state.artboards || [];
    layout(list);

    var seen = new Set();
    var changedID = null;
    var added = 0;

    list.forEach(function (data) {
      seen.add(data.id);
      var entry = boards.get(data.id);
      if (!entry) {
        entry = buildBoard(data);
        boards.set(data.id, entry);
        paintBoard(entry);
        log(data.title || data.slug || data.id, 'added');
        changedID = data.id;
        added++;
        return;
      }
      var before = entry.data;
      entry.data = data;
      if (!sameShape(before, data)) {
        paintBoard(entry);
        if (before.revision !== data.revision) {
          flash(entry);
          log(data.title || data.slug || data.id, 'updated');
          changedID = data.id;
        } else if (before.status !== data.status) {
          log(data.title || data.slug || data.id, data.status || 'ready');
          changedID = data.id;
        }
      } else {
        // Position can move without any visible property changing, when a
        // board ahead of this one in the row appeared or grew.
        paintBoard(entry);
      }
    });

    boards.forEach(function (entry, id) {
      if (seen.has(id)) return;
      entry.el.remove();
      boards.delete(id);
      log(entry.data.title || id, 'removed');
    });

    empty.hidden = boards.size > 0;
    refreshJump(list);

    if (boards.size === 0) return;
    if (!userPlacedView) {
      fitAll();
    } else if (follow && changedID) {
      focusBoard(changedID, 60);
    } else if (added > 0 && !follow) {
      // A new board may have landed outside the current view; say so rather
      // than moving a viewport the user placed deliberately.
      log('canvas', added + ' new artboard' + (added === 1 ? '' : 's') + ' — press F to fit');
    }
  }

  function refreshJump(list) {
    var current = jump.value;
    jump.textContent = '';
    var head = document.createElement('option');
    head.value = '';
    head.textContent = list.length ? 'Artboards (' + list.length + ')…' : 'Artboards…';
    jump.appendChild(head);
    list.forEach(function (data) {
      var opt = document.createElement('option');
      opt.value = data.id;
      opt.textContent = data.title || data.slug || data.id;
      jump.appendChild(opt);
    });
    if (list.some(function (d) { return d.id === current; })) jump.value = current;
  }

  function log(subject, what) {
    var li = document.createElement('li');
    var time = document.createElement('time');
    var now = new Date();
    time.textContent = String(now.getHours()).padStart(2, '0') + ':' +
      String(now.getMinutes()).padStart(2, '0') + ':' +
      String(now.getSeconds()).padStart(2, '0');
    var text = document.createElement('span');
    text.textContent = subject + ' — ' + what;
    li.appendChild(time);
    li.appendChild(text);
    activityList.insertBefore(li, activityList.firstChild);
    while (activityList.childElementCount > 40) {
      activityList.removeChild(activityList.lastElementChild);
    }
  }

  function setLive(kind, text) {
    dot.className = kind;
    liveText.textContent = text;
  }

  function poll() {
    fetch('artboards', { cache: 'no-store' })
      .then(function (response) {
        if (!response.ok) throw new Error('HTTP ' + response.status);
        return response.json();
      })
      .then(function (state) {
        failures = 0;
        setLive('', state.error ? 'live · ' + state.error : 'live');
        apply(state);
      })
      .catch(function (err) {
        failures++;
        // One missed poll is a hiccup; a run of them means the session behind
        // this canvas is gone, and the user should know why nothing moves.
        if (failures === 3) log('canvas', 'lost contact with Pando (' + err.message + ')');
        setLive(failures >= 3 ? 'down' : 'stale', failures >= 3 ? 'disconnected' : 'reconnecting…');
      });
  }

  // ------------------------------------------------------------ interaction

  function zoomAt(clientX, clientY, factor) {
    var rect = viewport.getBoundingClientRect();
    var px = clientX - rect.left;
    var py = clientY - rect.top;
    var next = Math.max(MIN_ZOOM, Math.min(MAX_ZOOM, view.k * factor));
    if (next === view.k) return;
    // Keep the world point under the cursor pinned to the cursor.
    view.x = px - (px - view.x) * (next / view.k);
    view.y = py - (py - view.y) * (next / view.k);
    view.k = next;
    userPlacedView = true;
    applyView();
  }

  viewport.addEventListener('wheel', function (event) {
    event.preventDefault();
    if (event.ctrlKey || event.metaKey) {
      zoomAt(event.clientX, event.clientY, Math.exp(-event.deltaY / 220));
      return;
    }
    // A plain trackpad scroll pans; a mouse wheel usually reports large deltaY
    // in whole lines, and zooming on it is what people expect from a canvas.
    if (event.deltaMode !== 0 || Math.abs(event.deltaY) > 60) {
      zoomAt(event.clientX, event.clientY, Math.exp(-event.deltaY / 500));
      return;
    }
    view.x -= event.deltaX;
    view.y -= event.deltaY;
    userPlacedView = true;
    applyView();
  }, { passive: false });

  var dragging = null;
  viewport.addEventListener('pointerdown', function (event) {
    if (event.button !== 0 && event.button !== 1) return;
    dragging = { x: event.clientX, y: event.clientY, vx: view.x, vy: view.y, id: event.pointerId };
    viewport.setPointerCapture(event.pointerId);
    viewport.classList.add('panning');
  });
  viewport.addEventListener('pointermove', function (event) {
    if (!dragging || dragging.id !== event.pointerId) return;
    view.x = dragging.vx + (event.clientX - dragging.x);
    view.y = dragging.vy + (event.clientY - dragging.y);
    userPlacedView = true;
    applyView();
  });
  function endDrag(event) {
    if (!dragging || dragging.id !== event.pointerId) return;
    viewport.releasePointerCapture(event.pointerId);
    viewport.classList.remove('panning');
    dragging = null;
  }
  viewport.addEventListener('pointerup', endDrag);
  viewport.addEventListener('pointercancel', endDrag);

  document.addEventListener('keydown', function (event) {
    if (event.target && /^(INPUT|SELECT|TEXTAREA)$/.test(event.target.tagName)) return;
    var step = event.shiftKey ? 240 : 80;
    switch (event.key) {
      case '+': case '=': zoomAt(window.innerWidth / 2, window.innerHeight / 2, 1.2); break;
      case '-': case '_': zoomAt(window.innerWidth / 2, window.innerHeight / 2, 1 / 1.2); break;
      case '0': view.k = 1; userPlacedView = true; applyView(); break;
      case 'f': case 'F': userPlacedView = false; fitAll(); break;
      case 'ArrowLeft': view.x += step; userPlacedView = true; applyView(); break;
      case 'ArrowRight': view.x -= step; userPlacedView = true; applyView(); break;
      case 'ArrowUp': view.y += step; userPlacedView = true; applyView(); break;
      case 'ArrowDown': view.y -= step; userPlacedView = true; applyView(); break;
      default: return;
    }
    event.preventDefault();
  });

  document.getElementById('zoom-in').addEventListener('click', function () {
    zoomAt(window.innerWidth / 2, window.innerHeight / 2, 1.2);
  });
  document.getElementById('zoom-out').addEventListener('click', function () {
    zoomAt(window.innerWidth / 2, window.innerHeight / 2, 1 / 1.2);
  });
  document.getElementById('fit').addEventListener('click', function () {
    userPlacedView = false;
    fitAll();
  });
  document.getElementById('actual').addEventListener('click', function () {
    view.k = 1;
    userPlacedView = true;
    applyView();
  });
  followBtn.addEventListener('click', function () {
    follow = !follow;
    followBtn.setAttribute('aria-pressed', String(follow));
  });
  jump.addEventListener('change', function () {
    if (!jump.value) return;
    userPlacedView = true;
    focusBoard(jump.value, 40);
  });
  window.addEventListener('resize', function () {
    if (!userPlacedView) fitAll();
  });

  applyView();
  poll();
  window.setInterval(poll, POLL_MS);
})();
